### PR TITLE
[sparse][ci] Add cuSPASRELt to CI

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115988
* __->__ #115369

Summary:

This PR adds in cuSPARSELt v0.4.07 into CI (12.1 and 11.8 CUDA) to run our cuSPARSELt specific tests.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: .ci/docker/common/install_cusparselt.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash
+
```
Comment (huydhn) [inline .ci/docker/common/install_cusparselt.sh:2]: I think you could add `set -ex` here to harden the script and also allow the bash steps to be printed on GitHub console log to very that the installation works correctly

File: .ci/docker/common/install_cusparselt.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/x86_64-linux-gnu/
+
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/local/cuda/lib64/
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/lib/x86_64-linux-gnu/
+    cd ..
+    rm -rf tmp_cusparselt
+    ldconfig
+fi
+
+echo "This is a test"
```
Comment (malfet) [inline .ci/docker/common/install_cusparselt.sh:30]:
```suggestion
set -ex

# cuSPARSELt license: https://docs.nvidia.com/cuda/cusparselt/license.html
mkdir tmp_cusparselt
pushd tmp_cusparselt
CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.4.0.7-archive"
curl --retry 3 -OLs "https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz"
tar xf "${CUSPARSELT_NAME}.tar.xz"
cp -a "${CUSPARSELT_NAME}/include/*" /usr/local/cuda/include/
cp -a ${CUSPARSELT_NAME}/lib/* /usr/local/cuda/lib64/
popd
rm -rf tmp_cusparselt
ldconfig
```

File: .ci/docker/common/install_cusparselt.sh [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+set -ex
+
+# cuSPARSELt license: https://docs.nvidia.com/cuda/cusparselt/license.html
+mkdir tmp_cusparselt
+pushd tmp_cusparselt
+CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.4.0.7-archive"
+curl --retry 3 -OLs "https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz"
+tar xf "${CUSPARSELT_NAME}.tar.xz"
+cp -a "${CUSPARSELT_NAME}/include/*" /usr/local/cuda/include/
```
Comment (jcaip) [inline .ci/docker/common/install_cusparselt.sh:10]: I think these quotes are breaking the script @malfet

File: test/test_sparse_semi_structured.py [hunk 4]
```diff
@@ -598,6 +598,8 @@ def setUp(self):
 
     @parametrize("dense_input_shape", [(128, 128)])
     def test_cslt_sparse_mm_int8_in_fp16_out(self, dense_input_shape, device):
+        assert False
```
Comment (jcaip) [inline test/test_sparse_semi_structured.py:601]: just to make sure it's getting built with cusparselt, will remove later

**Review by malfet — CHANGES_REQUESTED** [review]
[Edit] Please note, that `rebase` undid all the changes I've made yesterday, please get rid of  `USE_CUSPARSELT`  unless you plan to add CI config without it.
Also, don't plan to test two different versions of cusparseLT, unless there is a good reason to do it

File: test/test_sparse_semi_structured.py (hunk 4 above)
Comment (malfet) [reply test/test_sparse_semi_structured.py:601]: You were building with CUSPARELT, see https://github.com/pytorch/pytorch/actions/runs/7215814225/job/19660823130#step:12:1169

**Comment by jcaip** [comment]
@malfet Sorry, I rebased to fix tests, didn't realize it would overwrite your changes.

0.5.2.1 requires >12.0 and 0.4.0.7 requires > 11.8, which is why I added both. Should I just default to 0.5.2.1 and restrict it to run on the 12.1 runners only?

**Comment by malfet** [comment]
@jcaip can you share link to the doc that specifies min requirements for the cusparselt?

**Comment by jcaip** [comment]
https://docs.nvidia.com/cuda/cusparselt/release_notes.html

File: .ci/docker/common/install_cusparselt.sh [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+set -ex
+
+if [[ ${CUDA_VERSION:0:4} == "12.1" ]]; then
+    # cuSPARSELt license: https://docs.nvidia.com/cuda/cusparselt/license.html
+    mkdir tmp_cusparselt && cd tmp_cusparselt
+    CUSPARSELT_NAME="libcusparse_lt-linux-x86_64-0.5.2.1-archive"
+    curl --retry 3 -OLs https://developer.download.nvidia.com/compute/cusparselt/redist/libcusparse_lt/linux-x86_64/${CUSPARSELT_NAME}.tar.xz
+
+    tar xf ${CUSPARSELT_NAME}.tar.xz
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/
```
Comment (malfet) [inline .ci/docker/common/install_cusparselt.sh:12]:
```suggestion
```

File: .ci/docker/common/install_cusparselt.sh [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/local/cuda/include/
+    cp -a ${CUSPARSELT_NAME}/include/* /usr/include/x86_64-linux-gnu/
```
Comment (malfet) [inline .ci/docker/common/install_cusparselt.sh:14]:
```suggestion
```

File: .ci/docker/common/install_cusparselt.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/local/cuda/lib64/
+    cp -a ${CUSPARSELT_NAME}/lib/* /usr/lib/x86_64-linux-gnu/
```
Comment (malfet) [inline .ci/docker/common/install_cusparselt.sh:17]:
```suggestion
```

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**